src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -178,10 +178,50 @@ void Terminal::UpdateSettings(winrt::Microsoft::Terminal::Settings::ICoreSetting
 
     const short newBufferHeight = viewportSize.Y + _scrollbackLines;
     COORD bufferSize{ viewportSize.X, newBufferHeight };
-    RETURN_IF_FAILED(_buffer->ResizeTraditional(bufferSize));
 
-    auto proposedTop = oldTop;
-    const auto newView = Viewport::FromDimensions({ 0, proposedTop }, viewportSize);
+    // First allocate a new text buffer to take the place of the current one.
+    std::unique_ptr<TextBuffer> newTextBuffer;
+    try
+    {
+        newTextBuffer = std::make_unique<TextBuffer>(bufferSize,
+                                                     _buffer->GetCurrentAttributes(),
+                                                     0, // temporarily set size to 0 so it won't render.
+                                                     _buffer->GetRenderTarget());
+    }
+    CATCH_RETURN();
+
+    RETURN_IF_FAILED(TextBuffer::Reflow(*_buffer.get(), *newTextBuffer.get(), _mutableViewport));
+
+    // However conpty resizes a little oddly - if the height decreased, and
+    // there were blank lines at the bottom, those lines will get trimmed.
+    // If there's not blank lines, then the top will get "shifted down",
+    // moving the top line into scrollback.
+    // See GH#3490 for more details.
+
+    // If the final position in the buffer is on the bottom row of the new
+    // viewport, then we're going to need to move the top down. Otherwise,
+    // move the bottom up.
+    const auto dy = viewportSize.Y - oldDimensions.Y;
+    const COORD oldCursorPos = _buffer->GetCursor().GetPosition();
+
+#pragma warning(push)
+#pragma warning(disable : 26496) // cpp core checks wants this const, but it's assigned immediately below...

```

Comment:
` #pragma warning (suppress: 26496)`
should also work if you want a one liner instead of push/disable/pop.